### ISSUE TITLE
[XFCE] locked xscreensaver displays sensitive notifications

### ISSUE THREAD

**Issue by mfc** [QubesOS/qubes-issues#2026]
#### Qubes OS version (e.g., `R3.1`):

R3.1
#### Affected TemplateVMs (e.g., `fedora-23`, if applicable):

xscreensaver on XFCE 

---
### Expected behavior:

sensitive notifications are not displayed when the screen is locked.
### Actual behavior:

sensitive notifications are displayed when the screen is locked.
### Steps to reproduce the behavior:

Use Signal Desktop, receive a message. The message will be pushed through xscreensaver and display to the user even if they have locked the screen.

---
#### Related issues:

https://github.com/QubesOS/qubes-issues/issues/888
https://github.com/QubesOS/qubes-issues/issues/963
https://github.com/QubesOS/qubes-issues/issues/1917

**Comment by mfc** [comment]
just checked KDE screensaver and it does not leak these same notifications.

**Comment by mfc** [comment]
also a note that split-gpg notifications of private key access are pushed through the lock screen. This is on R3.2.

**Comment by andrewdavidwong** [comment]
This would be (hopefully) fixed by changing the default screen locker to physlock (#1917).

**Comment by cfcs** [comment]
@mfc KDE has the same problems, as outlined by @nvesely in https://github.com/QubesOS/qubes-issues/issues/963#issuecomment-185436718

(also, this bug is still a problem in Q3.2)

**Comment by andrewdavidwong** [comment]
I've found that this is not a Qubes-specific issue. I see popup notifications display through xscreensaver all the time on multiple physically distinct machines running an unrelated Linux distro. In light of this, @marmarek, should this issue be closed as `notourbug`?

**Comment by cfcs** [comment]
@andrewdavidwong It's a bug that we expose our users to by the choice of software we include, and by choice, since there are solutions available that does not have these security flaws (things leaking, bypassing authentication when the application crashes). Notably `physlock`, which we have been discussing for years in these Github threads now.

Saying that it's `notourbug` is IMHO equal to saying `meltdown / spectre` is not our bug, so lets ignore it, and if that is the strategy that Qubes wants to go with, we could erase most of the Qubes security bulletins already, since even if they affect Qubes users, very few of them are in software maintained by the Qubes team.

**Comment by andrewdavidwong** [comment]
> It's a bug that we expose our users to by the choice of software we include, and by choice, since there are solutions available that does not have these security flaws (things leaking, bypassing authentication when the application crashes). Notably `physlock`, which we have been discussing for years in these Github threads now.

It's not entirely "by choice." If we could wave a magic wand and switch to `physlock` with everything working perfectly, I'm sure we would've done so years ago. But switching is not easy, and doing so carries opportunity costs.

> Saying that it's `notourbug` is IMHO equal to saying `meltdown / spectre` is not our bug, so lets ignore it, 

No, it's not. Closing this issue as `notourbug` does not entail that we'll ignore the screeenlocker topic in general. To be precise, closing #2026 does not entail closing #1917. They're two distinct issues. If #2026 is genuinely a bug in an upstream project's software and not ours, and if our developers cannot or will not fix that bug for the other project, then closing #2026 as `notourbug` is probably the correct thing to do. It doesn't follow that we're going to ignore the underlying issue or that we don't care about it. On the contrary, if it looks like the upstream project isn't going to fix #2026 (and we can't or won't do it for them), that might be all the more reason to increase the priority of #1917.

> and if that is the strategy that Qubes wants to go with, we could erase most of the Qubes security bulletins already, since even if they affect Qubes users, very few of them are in software maintained by the Qubes team.

No. We issue QSBs to notify users of security-critical bugs and how to patch them (among other reasons). Given the nature of Qubes, most bugs (e.g., web browser bugs) are not security critical for Qubes, so we don't publish QSBs about them. As you know, most QSBs are about Xen bugs, because Xen is security critical. The fact that the Qubes team does not maintain Xen (the main project) is irrelevant to whether we issue QSBs for security-critical Xen bugs that affect Qubes.

**Comment by cfcs** [comment]
- Re: "by choice": Maybe that was a poor/provocative choice of words, I'm sorry. I appreciate that every reconfiguration carries cost, and it's a question of priorities. I personally think it's a somewhat important aspect of a distribution that aims to provide people with a secure desktop, but if a working screen locker is not a priority that the Qubes team shares, and the only security-critical component worth issuing QSBs about is in fact Xen, I feel it would be more honest to be up front about that, or just removing xscreensaver entirely and let people who cares about being able to safely lock their screen seek out their own solution.
At least this stance comes as a surprise to me, and I think it will be surprising to people who currently use Qubes in shared office spaces, in transit, or anywhere their equipment may conceivably be stolen or momentarily accessed by other people.

It is my understanding that your argument for refraining from noticing people about bugs in browsers, or the recent LibreOffice exploit (that still works?) is that they expose AppVMs, but not dom0, to danger, and you don't care about what runs in people's AppVMs. The screen locker runs in dom0 and protects the whole system, so I don't see how that argument intuitively extends to mean that screenlocking is not security-critical.

- Re: Is it it a bug? As [detailed by upstream in the FAQ](https://www.jwz.org/xscreensaver/faq.html#popup-windows) this is not so much a bug in XScreenSaver as it is a requirement that was overlooked in the design of X11 (and X.org) decades ago, and neglected in the favor of backwards compatibility ever since. This means #2026 **will not be fixed upstream** since it **cannot** be fixed.

XScreenSaver in particular has had a few "exciting" bugs that allowed bypassing the lock screen:
- [2015: Plugging in/out a HDMI cable](https://www.debian.org/security/2016/dsa-3438)
- [2012: Press Ctrl-Alt-*](https://bugs.archlinux.org/index.php?do=details&task_id=27993)

But this is not exclusive to XScreenSaver (see #963 for the exact same issue in KDE's screenlocker, also on Qubes; [and here @marmarek also reports that it's broken](https://github.com/QubesOS/qubes-issues/issues/888#issuecomment-77762358)); building an X11 screensaver application [is a design flaw in itself](https://github.com/QubesOS/qubes-issues/issues/963#issuecomment-185436718). Originally screensavers were meant to - save the screen from burning a permanent image, not to act as a security-critical component. The latter part was bolted onto the solutions as an afterthought, and it kind of shows.

- `notourbug`: When upstream can't fix a security flaw, and you won't fix it, closing the github issue that collects information about it removes visibility of the bug, which I feel is detrimental to the user experience. If a security-focused software distribution that I am using ships with known security flaws, I would like to know about them so that they can be weighed into my workflow. If people were made aware that their screen content would leak, they could choose not to use Qubes in settings where other people may see their (locked) screen (or install `physlock` or seek help to mitigate the problem themselves, like installing `physlock` or similar [using the instructions provided](https://github.com/QubesOS/qubes-issues/issues/963#issuecomment-107705261) in the github issue thread).

**Comment by cfcs** [comment]
BTW, found this other related thread (now closed for some reason) where @rootkovska and @marmarek also discuss screen locking: https://github.com/QubesOS/qubes-issues/issues/2120

**Comment by cfcs** [comment]
Was closing the tabs from yesterday when I spotted this quote from the author of XScreenSaver regarding his apparent frustration with the situation (which is that you cannot write a secure X11-based screensaver):
> There's little that I can do to make the screen locker secure so long as the kernel and X11 developers are actively working against security. The strength of the lock on your front door doesn't matter much so long as someone else in the house insists on leaving a key under the welcome mat.

I really don't want to sound too bitter or accusatory, please don't read any of this as an attack. I'm just a little weary of having discussed this in Github threads for 3 years now, and it seems like it is very unintuitive to people that `xscreensaver`, KDE's locker program etc are essentially **toys**, not security mechanisms.

**Comment by andrewdavidwong** [comment]
> I personally think it's a somewhat important aspect of a distribution that aims to provide people with a secure desktop, but if a working screen locker is not a priority that the Qubes team shares, and the only security-critical component worth issuing QSBs about is in fact Xen, I feel it would be more honest to be up front about that, or just removing xscreensaver entirely and let people who cares about being able to safely lock their screen seek out their own solution.

A secure screen locker is a high priority for us. We care about the security of Qubes OS as a whole, not just Xen. It just happens to be the case that many QSBs have been about Xen because many of the security-critical bugs that have affected Qubes have been in Xen.

> When upstream can't fix a security flaw, and you won't fix it, closing the github issue that collects information about it removes visibility of the bug, which I feel is detrimental to the user experience. If a security-focused software distribution that I am using ships with known security flaws, I would like to know about them so that they can be weighed into my workflow. If people were made aware that their screen content would leak, they could choose not to use Qubes in settings where other people may see their (locked) screen (or install physlock or seek help to mitigate the problem themselves, like installing physlock or similar using the instructions provided in the github issue thread).

It sounds like the proper place to communicate such information is in the documentation, release notes, or in the installer. GitHub issues are meant for tracking issues, not for notifying users about things. If the crux of this issue is that users need to be informed about it, then it should be turned into a documentation issue (or similar), and closed once the documentation (or announcement or whatever) is done. The fact that people should be aware of something is not a good argument for leaving a GitHub issue about that thing open, since many users never see the issue tracker, and those who do can easily search through both open and closed issues. More importantly, leaving issues open when they aren't actionable makes the entire issue tracking system less efficient, which hinders our developers from doing their work.

**Comment by cfcs** [comment]
@andrewdavidwong I think it would be very good if this was detailed in the documentation (I guess release notes are more for information specific to a certain release), and/or in the installer indeed.
But right now this is not mentioned *anywhere else,* which is why I'm concerned that by closing the issue we would lose the only place where users can currently learn about the issue.
Information dissemination aside, this **is** an _issue_ that affects current versions of Qubes, and therefore I do think it belongs in the *_issue_ tracker*. It is a technical issue to which concrete solutions have been proposed, so I would prefer the fix being implemented in Qubes rather than shipping affected releases and informing that user that it comes pre-installed with security vulnerabilities. The latter is a perfectly valid approach to dealing with security vulnerabilities that are impossible to solve, but where solutions are available I find it suboptimal.

**It is** an actionable issue; the fix is simple: Use physlock instead.

**Comment by andrewdavidwong** [comment]
> I think it would be very good if this was detailed in the documentation (I guess release notes are more for information specific to a certain release), and/or in the installer indeed.
But right now this is not mentioned *anywhere else,* which is why I'm concerned that by closing the issue we would lose the only place where users can currently learn about the issue.
Information dissemination aside, this **is** an _issue_ that affects current versions of Qubes, and therefore I do think it belongs in the *_issue_ tracker*.

Please read my comment above again. Specifically:

"If the crux of this issue is that users need to be informed about it, then it should be turned into a documentation issue (or similar), and closed once the documentation (or announcement or whatever) is done."

> It is a technical issue to which concrete solutions have been proposed, so I would prefer the fix being implemented in Qubes rather than shipping affected releases and informing that user that it comes pre-installed with security vulnerabilities. The latter is a perfectly valid approach to dealing with security vulnerabilities that are impossible to solve, but where solutions are available I find it suboptimal.
> 
> **It is** an actionable issue; the fix is simple: Use physlock instead.

That's precisely what #1917 is about.

**Comment by cfcs** [comment]
@andrewdavidwong Ah, sorry: **This** thread is about the glitching, and #1917 is about the fix.
Then I agree with you. :-)

**Comment by andrewdavidwong** [comment]
This issue is being closed because:

 - This issue has the "Release 3.2 updates" milestone.
 - [Qubes OS 3.2 recently reached end-of-life (EOL).](https://www.qubes-os.org/news/2019/03/28/qubes-3-2-has-reached-eol/)
 - This issue has not been updated in over a year.

If anyone believes that this issue should be reopened, please let us know in a comment here.

**Comment by iamahuman** [comment]
This issue should be reopened since:

 - It persists into Qubes OS 4.0.
 - It probably also persists into Qubes OS 4.1.
 - Opening a new issue seems unnecessary since there is nothing new to add to the problem.

**Comment by iamahuman** [comment]
For users that considers this issue to be a very serious security problem and don't need pop-up windows, a temporary workaround is to set `override_redirect_protection = true;`, introduced in [this PR](https://github.com/QubesOS/qubes-gui-daemon/pull/41) which actually aims to fix #5731.

**Comment by marmarek** [comment]
>  set `override_redirect_protection = true;`

This a) is the default value; b) won't help, as it applies only to large windows (>90% screen size).

**Comment by marmarek** [comment]
Regarding specifically notification, #889 should solve this too, because notifications would be displayed directly by an application in dom0/GUI domain, instead of from within a VM (which don't know anything about screenlocker).

**Comment by tlaurion** [comment]
@marmarek 
Riot and Signal notifications are now appearing over screensaver, disclosing sensitive information when screen is locked.

**Comment by marmarek** [comment]
Xscreensaver info on this: https://www.jwz.org/xscreensaver/faq.html#popup-windows
Doesn't look very optimistic.

**Comment by andrewdavidwong** [comment]
> Xscreensaver info on this: https://www.jwz.org/xscreensaver/faq.html#popup-windows
> Doesn't look very optimistic.

Sounds like it's `not our bug`. Do we still wish to keep this issue open?

**Comment by h01ger** [comment]
On Fri, Jul 10, 2020 at 10:47:42PM -0700, Andrew David Wong wrote:
> > Xscreensaver info on this: https://www.jwz.org/xscreensaver/faq.html#popup-windows
> > Doesn't look very optimistic.
> Sounds like it's `not our bug`. Do we still wish to keep this issue open?

a fix would be installing and using physlock instead of xscreensaver...


-- 
cheers,
	Holger

-------------------------------------------------------------------------------
               holger@(debian|reproducible-builds|layer-acht).org
       PGP fingerprint: B8BF 5413 7B09 D35C F026 FE9D 091A B856 069A AA1C

**Comment by mfc** [comment]
> > Xscreensaver info on this: https://www.jwz.org/xscreensaver/faq.html#popup-windows
> > Doesn't look very optimistic.
> 
> Sounds like it's `not our bug`. Do we still wish to keep this issue open?

it is not our bug but it negatively affects the security & privacy of Qubes users. so we need to address it in some way -- either by our choice of screensaver, mitigating in some way, or fixing upstream.

**Comment by andrewdavidwong** [comment]
> > > Xscreensaver info on this: https://www.jwz.org/xscreensaver/faq.html#popup-windows
> > > Doesn't look very optimistic.
> > 
> > 
> > Sounds like it's `not our bug`. Do we still wish to keep this issue open?
> 
> it is not our bug but it negatively affects the security & privacy of Qubes users. so we need to address it in some way -- either by our [choice of screensaver](https://github.com/QubesOS/qubes-issues/issues/1917), [mitigating in some way](https://github.com/QubesOS/qubes-issues/issues/889), or fixing upstream.

You are mistaking mundane issue tracker management for high-level decision-making, willingness, or commitment. They are separate matters. In particular, closing an issue as `not our bug` in this issue tracker has a very specific meaning: that the particular bug reported in the issue is not a bug in something controlled by the Qubes OS Project.

If we were to keep an issue open simply to signal our willingness or commitment to some cause when we know that specific issue is not actionable (rather than a related issue or one in an upstream project's issue tracker), it would dilute the usefulness of the issue tracker as a technical tool for the developers who rely on it for their daily work.

As you yourself point out, there are two related issues open, and one or both of those might be the way in which we address the broader problem here. Therefore, closing this particular issue does not mean we won't address the broader problem.

Do not make the mistake of thinking "not our bug" means "not our problem." No one has ever said that. The Qubes developers should never be *expected* to fix bugs in code from upstream projects, though many times do. But those are typically tracked in the upstream project's issue tracker rather than ours, because that makes much more organizational sense for everyone involved.

Examples of good reasons to leave this open:
 - We discover it's actually our bug after all
 - We change the issue to be mean something else (i.e., change it from no longer reporting an upstream bug to being something actionable by Qubes devs)
 - We can implement some kind of fix for the upstream bug on our end
 - The Qubes devs want to use it to keep track of when it gets fixed upstream
 - Leaving it open is useful for the Qubes devs in some other way

Examples of bad reasons to leave this open:
 - We decide it's not actionable, but we want to show we care.
 - We decide it's not actionable, but it's related to the other open issues about this that are actually actionable.
 - We decide it's not actionable, but people are still affected by the problem. (An unactionable issue doesn't help them and could actually hurt. If a zombie issue like that remains open, people might mistakenly assume that we have a plan for how to proceed or that it's being worked on, for example. Or they might get angry that nothing has been done with it after a long time.)

**Comment by cfcs** [comment]
My point of view is that his is not a bug in xscreensaver, it's not a bug in QubeOS, but it *is a security issue* in QubesOS as a consequence of choosing a software solution to a problem (screenlocking) that does not aim to match the expectations of QubesOS users (as demonstrated by the barrage of users reporting this as a problem).

- Which is namely that the screenlocker will a) protect them from passers-by interacting with their desktop b) protect them from passers-by looking at private information on the desktop while locked. Xscreensaver casually aims to provide feature a), but the documentation also clearly states that it's a "best-effort" kind of solution with multiple known failure modes that leaves the desktop unprotected. It also clearly states that b) is out of scope.

There are several ways to solve this situation, one would be to document somewhere that QubesOS users should not harbor these expectations, and that if they want a proper screen locking mechanism they could consider running an operating system that provides a level of security adequate to their demands (like Windows or MacOS), or they could follow the instructions for installing e.g. `physlock` as detailed in a number of the many threads about this issue that live here on the QubesOS issue tracker.

That being said, now we have been bickering forth and back about this issue across multiple issue threads for the last 4-5 years.

About every 3-6 months someone stops by to leave a screenshot of this perceived security problem in QubesOS.
I'd be for closing or at the very least locking all the issue threads related to this, since it's annoying to keep getting notifications about this when nobody wants to change it.

**Comment by h01ger** [comment]
On Tue, Jul 14, 2020 at 02:13:05PM -0700, cfcs wrote:
> My point of view is that his is not a bug in xscreensaver, it's not a bug in QubeOS, but it *is a security issue* in QubesOS as a consequence of choosing a software solution to a problem (screenlocking) that does not aim to match the expectations of QubesOS users (as demonstrated by the barrage of users reporting this as a problem).

agreed and thanks for summarizing!

> There are several ways to solve this situation, one would be to document somewhere that QubesOS users should not harbor these expectations, and that if they want a proper screen locking mechanism they could consider running an operating system that provides a level of security adequate to their demands (like Windows or MacOS), or they could follow the instructions for installing e.g. `physlock` as detailed in a number of the many threads about this issue that live here on the QubesOS issue tracker.

@cfcs: you are not saying it directly, but would you agree that switching to
physlock would be the technical solution you'd prefer? From what I've read, I
agree. I just now wonder about wayland, though IIRC physlock would also work
with that.

I'd definitly like QubesOS to be "secure" and ahead of other OS in this aspect
as well! and physlock seems to be *the* default for *this* aspect of OS! :)


-- 
cheers,
	Holger

-------------------------------------------------------------------------------
               holger@(debian|reproducible-builds|layer-acht).org
       PGP fingerprint: B8BF 5413 7B09 D35C F026 FE9D 091A B856 069A AA1C

**Comment by mfc** [comment]
thanks for the clear summary @cfcs and for the idea of locking the thread which hadn't been discussed previously.

issues like #889 or #1917 are framed as enhancements which misses that they are also the (potential) mitigations of this Qubes OS security issue. because of this, they may not be properly prioritized without this issue continuing to be present for context.

i would propose keeping this issue open and locking the thread, and we can close it once #889 or #1917 is implemented and we confirm that the info leak no longer occurs.

**Comment by andrewdavidwong** [comment]
> thanks for the clear summary @cfcs and for the idea of locking the thread which hadn't been discussed previously.
> 
> issues like #889 or #1917 are framed as enhancements which misses that they are also the (potential) mitigations of this Qubes OS security issue. because of this, they may not be properly prioritized without this issue continuing to be present for context.

#889 and #1917 each have their own priority levels, which you can see on their labels. If we want to state *why* they have the priority they do, we do that on each respective issue, and we link to this issue to provide context and justification. I'll do that on #1917 now. None of that requires this issue to remain open. This issue continues to be present even if it's closed.

> i would propose keeping this issue open and locking the thread, and we can close it once #889 or #1917 is implemented and we confirm that the info leak no longer occurs.

This should be closed as `not our bug` (as long as that's accurate). There is generally no reason to keep a defunct issue open until a separate one is implemented. That would suggest that the issues are inadequately organized or defined. We should shore that up by fleshing out #889 and #1917, if needed. If there's still a leak in physlock, that would be a separate bug report.

I do, however, agree with locking this issue.

No, we should close this as `not our bug`.

**Comment by marmarek** [comment]
While issue in xscreensaver indeed is something we won't solve in xscreensaver directly, there are a bunch of alternative solutions.
I would like to add another one to the list: #5908 - this one (contrary to physlock or light-locker) does not involve switching VT which is known to cause issues with some graphics drivers.
